### PR TITLE
[ISSUE-883] Transitively export the dependencies of [eventmesh-sdk-java] to other module

### PR DESCRIPTION
**Pull request by HoffmanZheng**
Fixes ISSUE #883 .

### Motivation

[eventmesh-sdj-java](https://mvnrepository.com/artifact/org.apache.eventmesh/eventmesh-sdk-java) is a library module, should transitively export the dependencies, which are necessary for development of [eventmesh-application](https://github.com/apache/incubator-eventmesh/tree/master/eventmesh-examples).

### Modifications

Add gradle plugin [java-library](https://docs.gradle.org/current/userguide/java_library_plugin.html) in eventmesh-sdk-java

Change the configuration of following dependencies from `implementation` to `api`: 
* project(":eventmesh-common")
* io.cloudevents:cloudevents-core
* io.cloudevents:cloudevents-json-jackson
* io.openmessaging:openmessaging-api
* org.projectlombok:lombok:1.18.22

### Documentation

Not a new feature, thus no doc needed.

### CODE REVIEW

**Review by HoffmanZheng — COMMENTED** [review]
Though [maven-publish](https://github.com/apache/incubator-eventmesh/blob/53c62509012dafa4d0912f9f7cb35d5be6c11389/build.gradle#L52) is included in the project, there should be a committor to do the related test.

File: eventmesh-sdk-java/build.gradle [hunk 1]
```diff
@@ -15,10 +15,15 @@
  * limitations under the License.
  */
 
+plugins {
+    id 'java-library'
+    id 'com.github.johnrengelman.shadow' version '7.1.2'
```
Comment (HoffmanZheng) [inline eventmesh-sdk-java/build.gradle:20]: After I use the `java-library` and change the dependency configuration to `api`, the built jar file still does not contain the dependencies.
Thus the `shadow` plugin used here to generate a fat jar. (for your [reference](https://discuss.gradle.org/t/how-to-include-dependencies-in-jar/19571))

**Review by HoffmanZheng — COMMENTED** [review]
Though [maven-publish](https://github.com/apache/incubator-eventmesh/blob/53c62509012dafa4d0912f9f7cb35d5be6c11389/build.gradle#L52) is included in the project, there should be a committor to do the related test.

**Comment by ruanwenjun** [comment]
Good catch. Right now, the scope of `eventmesh-common` in `eventmesh-sdk-java` is `implementation`, so if we want to use eventmesh-sdk-java in our app, we need to add the `eventmesh-common` to our dependency. We need to change the scope from `implementation` to api to solve this.

And I think we don't need to make a shawdow jar. cc @xwm1992

**Comment by HoffmanZheng** [comment]
> Good catch. Right now, the scope of `eventmesh-common` in `eventmesh-sdk-java` is `implementation`, so if we want to use eventmesh-sdk-java in our app, we need to add the `eventmesh-common` to our dependency. We need to change the scope from `implementation` to api to solve this.
> 
> And I think we don't need to make a shawdow jar. cc @xwm1992

Currently I'm trying to use the following script to generate a fat jar of [eventmesh-sdk-java], sometimes it works. Mark the PR work in progress.

```
jar {
    from {
        configurations. runtimeClasspath.filter{ it.exists() }.collect { it.isDirectory() ? it : zipTree(it) }
    }
}
```

**Comment by ruanwenjun** [comment]
> > Good catch. Right now, the scope of `eventmesh-common` in `eventmesh-sdk-java` is `implementation`, so if we want to use eventmesh-sdk-java in our app, we need to add the `eventmesh-common` to our dependency. We need to change the scope from `implementation` to api to solve this.
> > And I think we don't need to make a shawdow jar. cc @xwm1992
> 
> Currently I'm trying to use the following script to generate a fat jar of [eventmesh-sdk-java] (without the usage of `shadow` plugin), sometimes it works. Mark the PR work in progress.
> 
> ```
> jar {
>     from {
>         configurations. runtimeClasspath.filter{ it.exists() }.collect { it.isDirectory() ? it : zipTree(it) }
>     }
> }
> ```

I suggest we just change the scope of `eventmesh-commom` to `spi` is enough.

**Comment by HoffmanZheng** [comment]
I used to use the `gradle jar` and `mvn install:install-file` to do dependencies resolving test, and it does not work.
Rather using `gradle publishToMavenLocal`, the dependencies could be transitively exported to other module!

File: eventmesh-sdk-java/build.gradle [hunk 2]
```diff
@@ -15,10 +15,14 @@
  * limitations under the License.
  */
 
+plugins {
+    id 'java-library'
+}
+
```
Comment (ruanwenjun) [inline eventmesh-sdk-java/build.gradle:21]: It seems this already defined at root `build.gradle`
```suggestion
```

**Review by ruanwenjun — CHANGES_REQUESTED** [review]

**Comment by qqeasonchen** [comment]
@HoffmanZheng any update？

File: eventmesh-sdk-java/build.gradle [hunk 3]
```diff
@@ -50,10 +54,9 @@ dependencies {
     implementation "io.grpc:grpc-protobuf:1.15.0"
     implementation "io.grpc:grpc-stub:1.15.0"
     implementation "com.google.protobuf:protobuf-java-util:3.5.1"
-    compileOnly 'org.projectlombok:lombok:1.18.22'
+    api 'org.projectlombok:lombok:1.18.22'
```
[the remark this answers is not in the document]
Comment (HoffmanZheng) [reply eventmesh-sdk-java/build.gradle:53]: also make sense to me, seems the `eventmesh-example` uses that only for log. it's not necessary to export to the users of `eventmesh-sdk-java`
https://github.com/apache/incubator-eventmesh/blob/c055e8324fd6b0c67055b2f4c6731f386acdb5a4/eventmesh-examples/src/main/java/org/apache/eventmesh/http/demo/sub/controller/SubController.java#L42
https://github.com/apache/incubator-eventmesh/blob/c055e8324fd6b0c67055b2f4c6731f386acdb5a4/eventmesh-examples/src/main/java/org/apache/eventmesh/http/demo/sub/controller/SubController.java#L56

**Comment by HoffmanZheng** [comment]
> @HoffmanZheng any update？

updated

**Review by ruanwenjun — APPROVED** [review]
LGTM

**Pull request merged**